File: lint-rules-android-lint/src/test/java/com/vanniktech/lintrules/android/InvalidStringDetectorTest.java
package com.vanniktech.lintrules.android;

import com.android.tools.lint.checks.infrastructure.LintDetectorTest;
import com.android.tools.lint.detector.api.Detector;
import com.android.tools.lint.detector.api.Issue;
import java.util.List;
import org.intellij.lang.annotations.Language;

import static com.vanniktech.lintrules.android.AndroidDetectorTest.NO_WARNINGS;
import static com.vanniktech.lintrules.android.InvalidStringDetector.ISSUE_INVALID_STRING;
import static java.util.Collections.singletonList;
import static org.fest.assertions.api.Assertions.assertThat;

public class InvalidStringDetectorTest extends LintDetectorTest {
  public void testValidString() throws Exception {
    @Language("XML") final String source = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
        + "<resources>\n"
        + "  <string name=\"my_string\">My string</string>\n"
        + "</resources>";

    assertThat(lintProject(xml("/res/values/strings.xml", source))).isEqualTo(NO_WARNINGS);
  }

  public void testStringContainingNewLine() throws Exception {
    @Language("XML") final String source = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
        + "<resources>\n"
        + "  <string name=\"my_string\">My string\n"
        + "</string>\n"
        + "</resources>";

    assertThat(lintProject(xml("/res/values/strings.xml", source))).isEqualTo("res/values/strings.xml:3: Warning: Text contains new line. [InvalidString]\n"
        + "  <string name=\"my_string\">My string\n"
        + "                           ^\n"
        + "0 errors, 1 warnings\n");
  }

  public void testTrailingWhitespaceAtEndString() throws Exception {
    @Language("XML") final String source = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
        + "<resources>\n"
        + "  <string name=\"my_string\">My string   </string>\n"
        + "</resources>";

    assertThat(lintProject(xml("/res/values/strings.xml", source))).isEqualTo("res/values/strings.xml:3: Warning: Text contains trailing whitespace. [InvalidString]\n"
        + "  <string name=\"my_string\">My string   </string>\n"
        + "                           ^\n"
        + "0 errors, 1 warnings\n");
  }

  public void testTrailingWhitespaceAtBeginningString() throws Exception {
    @Language("XML") final String source = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
        + "<resources>\n"
        + "  <string name=\"my_string\">  My string</string>\n"
        + "</resources>";

    assertThat(lintProject(xml("/res/values/strings.xml", source))).isEqualTo("res/values/strings.xml:3: Warning: Text contains trailing whitespace. [InvalidString]\n"
        + "  <string name=\"my_string\">  My string</string>\n"
        + "                             ^\n"
        + "0 errors, 1 warnings\n");
  }

  public void testTrailingWhitespacePluralString() throws Exception {
    @Language("XML") final String source = "<plurals name=\"days\">\n"
        + "    <item quantity=\"one\">%d Day</item>\n"
        + "    <item quantity=\"other\">%d Days</item>\n"
        + "  </plurals>";

    assertThat(lintProject(xml("/res/values/strings.xml", source))).isEqualTo(NO_WARNINGS);
  }

  public void testTrailingWhitespacePluralStringTrailing() throws Exception {
    @Language("XML") final String source = "<plurals name=\"days\">\n"
        + "    <item quantity=\"one\">  %d Day</item>\n"
        + "    <item quantity=\"other\">%d Days   </item>\n"
        + "  </plurals>";

    assertThat(lintProject(xml("/res/values/strings.xml", source))).isEqualTo("res/values/strings.xml:2: Warning: Text contains trailing whitespace. [InvalidString]\n"
        + "    <item quantity=\"one\">  %d Day</item>\n"
        + "                           ^\n"
        + "res/values/strings.xml:3: Warning: Text contains trailing whitespace. [InvalidString]\n"
        + "    <item quantity=\"other\">%d Days   </item>\n"
        + "                           ^\n"
        + "0 errors, 2 warnings\n");
  }

  @Override protected Detector getDetector() {
    return new InvalidStringDetector();
  }

  @Override protected List<Issue> getIssues() {
    return singletonList(ISSUE_INVALID_STRING);
  }

  @Override protected boolean allowCompilationErrors() {
    return false;
  }
}

File: lint-rules-android-lint/src/test/java/com/vanniktech/lintrules/android/InvalidStringDetectorTest.kt
package com.vanniktech.lintrules.android

import com.android.tools.lint.checks.infrastructure.TestFiles.xml
import com.android.tools.lint.checks.infrastructure.TestLintTask.lint
import org.junit.Test

class InvalidStringDetectorTest {
  @Test fun validString() {
    lint()
        .files(xml("res/values/strings.xml", """
          |<resources>
          |  <string name="my_string">My string</string>
          |</resources>""".trimMargin()))
        .issues(ISSUE_INVALID_STRING)
        .run()
        .expectClean()
  }

  @Test fun stringContainingNewLine() {
    lint()
        .files(xml("res/values/strings.xml", """
          |<resources>
          |  <string name="my_string">My string"
          |</string>
          |</resources>""".trimMargin()))
        .issues(ISSUE_INVALID_STRING)
        .run()
        .expect("""
          |res/values/strings.xml:2: Warning: Text contains new line. [InvalidString]
          |  <string name="my_string">My string"
          |  ^
          |0 errors, 1 warnings""".trimMargin())
        .expectFixDiffs("""
          |Fix for res/values/strings.xml line 1: Fix it:
          |@@ -2 +2
          |-   <string name="my_string">My string"
          |- </string>
          |+   <string name="my_string">My string"</string>
          |""".trimMargin())
  }

  @Test fun trailingWhitespaceAtEndString() {
    lint()
        .files(xml("res/values/strings.xml", """
          |<resources>
          |  <string name="my_string">My string   </string>
          |</resources>""".trimMargin()))
        .issues(ISSUE_INVALID_STRING)
        .run()
        .expect("""
          |res/values/strings.xml:2: Warning: Text contains trailing whitespace. [InvalidString]
          |  <string name="my_string">My string   </string>
          |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          |0 errors, 1 warnings""".trimMargin())
        .expectFixDiffs("""
          |Fix for res/values/strings.xml line 1: Fix it:
          |@@ -2 +2
          |-   <string name="my_string">My string   </string>
          |+   <string name="my_string">My string</string>
          |""".trimMargin())
  }

  @Test fun trailingWhitespaceAtBeginningOfString() {
    lint()
        .files(xml("res/values/strings.xml", """
          |<resources>
          |  <string name="my_string">  My string</string>
          |</resources>""".trimMargin()))
        .issues(ISSUE_INVALID_STRING)
        .run()
        .expect("""
          |res/values/strings.xml:2: Warning: Text contains trailing whitespace. [InvalidString]
          |  <string name="my_string">  My string</string>
          |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          |0 errors, 1 warnings""".trimMargin())
        .expectFixDiffs("""
          |Fix for res/values/strings.xml line 1: Fix it:
          |@@ -2 +2
          |-   <string name="my_string">  My string</string>
          |+   <string name="my_string">My string</string>
          |""".trimMargin())
  }

  @Test fun validPluralString() {
    lint()
        .files(xml("res/values/strings.xml", """
          |<plurals name="days">
          |  <item quantity="one">%d Day</item>
          |  <item quantity="other">%d Days</item>
          |</plurals>""".trimMargin()))
        .issues(ISSUE_INVALID_STRING)
        .run()
        .expectClean()
  }

  @Test fun trailingWhitespaceAtPluralStrings() {
    lint()
        .files(xml("res/values/strings.xml", """
          |<plurals name="days">
          |  <item quantity="one">  %d Day</item>
          |  <item quantity="other">%d Days   </item>
          |</plurals>""".trimMargin()))
        .issues(ISSUE_INVALID_STRING)
        .run()
        .expect("""
          |res/values/strings.xml:2: Warning: Text contains trailing whitespace. [InvalidString]
          |  <item quantity="one">  %d Day</item>
          |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          |res/values/strings.xml:3: Warning: Text contains trailing whitespace. [InvalidString]
          |  <item quantity="other">%d Days   </item>
          |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          |0 errors, 2 warnings""".trimMargin())
        .expectFixDiffs("""
          |Fix for res/values/strings.xml line 1: Fix it:
          |@@ -2 +2
          |-   <item quantity="one">  %d Day</item>
          |+   <item quantity="one">%d Day</item>
          |Fix for res/values/strings.xml line 2: Fix it:
          |@@ -3 +3
          |-   <item quantity="other">%d Days   </item>
          |+   <item quantity="other">%d Days</item>
          |""".trimMargin())
  }

  @Test fun validStringArrayString() {
    lint()
        .files(xml("res/values/strings.xml", """
          |<string-array name="foo">
          |  <item>1</item>
          |  <item>2</item>
          |</string-array>""".trimMargin()))
        .issues(ISSUE_INVALID_STRING)
        .run()
        .expectClean()
  }

  @Test fun trailingWhitespaceAtStringArrayStrings() {
    lint()
        .files(xml("res/values/strings.xml", """
          |<string-array name="bar">
          |  <item>   1</item>
          |  <item>2   </item>
          |</string-array>""".trimMargin()))
        .issues(ISSUE_INVALID_STRING)
        .run()
        .expect("""
          |res/values/strings.xml:2: Warning: Text contains trailing whitespace. [InvalidString]
          |  <item>   1</item>
          |  ~~~~~~~~~~~~~~~~~
          |res/values/strings.xml:3: Warning: Text contains trailing whitespace. [InvalidString]
          |  <item>2   </item>
          |  ~~~~~~~~~~~~~~~~~
          |0 errors, 2 warnings""".trimMargin())
        .expectFixDiffs("""
          |Fix for res/values/strings.xml line 1: Fix it:
          |@@ -2 +2
          |-   <item>   1</item>
          |+   <item>1</item>
          |Fix for res/values/strings.xml line 2: Fix it:
          |@@ -3 +3
          |-   <item>2   </item>
          |+   <item>2</item>
          |""".trimMargin())
  }
}
